brave/src/main/java/brave/propagation/CompositePropagation.java

```diff
@@ -0,0 +1,144 @@
+ * successful, at which point the remaining {@link Propagation}s will not be called. If all {@link Propagation}s fail
+ * to extract, {@link TraceContextOrSamplingFlags#EMPTY} is returned.
+ */
+public class CompositePropagation<K> implements Propagation<K> {
+  public static FactoryBuilder newFactoryBuilder() {
+    return new FactoryBuilder();
+  }
+
+  /**
+   * Defaults to {@code injectAll = true}.
+   */
+  public static final class FactoryBuilder {
+    boolean injectAll = true;
+    final List<Propagation.Factory> propagationFactories = new ArrayList<>();
+
+    public FactoryBuilder addPropagationFactory(final Propagation.Factory propagationFactory) {
+      if (propagationFactory == null) throw new NullPointerException("propagationFactory == null");
+      propagationFactories.add(propagationFactory);
+      return this;
+    }
+
+    public FactoryBuilder addAllPropagationFactories(final Collection<Propagation.Factory> propagationFactories) {
+      if (propagationFactories == null) throw new NullPointerException("propagationFactories == null");
+      this.propagationFactories.addAll(propagationFactories);
+      return this;
+    }
+
+    public FactoryBuilder clearPropagationFactories() {
+      propagationFactories.clear();
+      return this;
+    }
+
+    public FactoryBuilder injectAll(final boolean injectAll) {
+      this.injectAll = injectAll;
+      return this;
+    }
+
+    public Propagation.Factory build() {
+      return new Factory(this);
+    }
+  }
+
+  private final List<Propagation<K>> propagations;
+  private final List<K> keys;
+  private final boolean injectAll;
+
+  CompositePropagation(final List<Propagation<K>> propagations, final boolean injectAll) {
+    this.propagations = propagations;
+    this.injectAll = injectAll;
+    final Set<K> keySet = new LinkedHashSet<>();
+    for (final Propagation<K> propagation : propagations) {
+      keySet.addAll(propagation.keys());
+    }
+    keys = new ArrayList<>(keySet);
+  }
+
+  @Override public List<K> keys() {
+    return keys;
+  }
+
+  @Override public <C> TraceContext.Injector<C> injector(final Setter<C, K> setter) {
+    return new TraceContext.Injector<C>() {
+      @Override
+      public void inject(final TraceContext traceContext, final C carrier) {
+        for (final Propagation<K> propagation : propagations) {
+          propagation.injector(setter).inject(traceContext, carrier);
+          if (!injectAll) {
+            break;
+          }
+        }
+      }
+    };
+  }
+
+  @Override public <C> TraceContext.Extractor<C> extractor(final Getter<C, K> getter) {
+    return new TraceContext.Extractor<C>() {
+      @Override
+      public TraceContextOrSamplingFlags extract(final C carrier) {
+        for (final Propagation<K> propagation : propagations) {
+          final TraceContextOrSamplingFlags result = propagation.extractor(getter).extract(carrier);
+          if (SamplingFlags.EMPTY != result.samplingFlags()) {
```

Comment:
sorry this isn't quite right, as when extra field/baggage is in use it won't necessarily equal the constant.

Reply:
I made an update that I think fixes this issue